File: Java/JavaServerDemo/src/com/topflytech/demo/MqttExample.java
package com.topflytech.demo;

import com.topflytech.codec.*;
import com.topflytech.codec.entities.*;
import io.netty.channel.Channel;

import java.io.IOException;
import java.util.List;

public class MqttExample {

    private String noObdImei = "518000518001517";
    private String obdImei = "518000518001518";
    private String personalObjImei = "518000518001519";
    private String mqttUrl = "tcp://192.168.1.251:9083";
    private SelfMqttClient noObdClient,obdClient,personalClient;
    private T880xdEncoder t880xdEncoder = new T880xdEncoder(MessageEncryptType.NONE,null);
    Decoder decoder = new Decoder(MessageEncryptType.NONE,null);
    private T880xPlusEncoder t880xPlusEncoder = new T880xPlusEncoder(MessageEncryptType.NONE,null);
    com.topflytech.codec.ObdDecoder obdDecoder = new com.topflytech.codec.ObdDecoder(MessageEncryptType.NONE,null);
    private PersonalAssetMsgEncoder personalAssetMsgEncoder = new PersonalAssetMsgEncoder(MessageEncryptType.NONE,null);
    PersonalAssetMsgDecoder personalAssetDecoder = new com.topflytech.codec.PersonalAssetMsgDecoder(MessageEncryptType.NONE,null);
    public void doMqttTest(){
        noObdClient = new SelfMqttClient(mqttUrl, noObdImei, new SelfMqttClient.MqttMsgCallback() {
            public void onReceiveMsg(byte[] hexBytes) {
                List<Message> messageList = decoder.decode(hexBytes);
                for (Message msg :
                        messageList) {
                    try {
                        dealNoObdMessage(msg);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
        noObdClient.connect();
        obdClient = new SelfMqttClient(mqttUrl, obdImei, new SelfMqttClient.MqttMsgCallback() {
            public void onReceiveMsg(byte[] hexBytes) {
                List<Message> messageList = obdDecoder.decode(hexBytes);
                for (Message msg :
                        messageList) {
                    try {
                        dealObdMessage(msg);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
        obdClient.connect();
        personalClient = new SelfMqttClient(mqttUrl, personalObjImei, new SelfMqttClient.MqttMsgCallback() {
            public void onReceiveMsg(byte[] hexBytes) {
                List<Message> messageList = personalAssetDecoder.decode(hexBytes);
                for (Message msg :
                        messageList) {
                    try {
                        dealPersonalAssetMessage(msg);
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
        personalClient.connect();
    }

    public void stopTest(){
        if(noObdClient != null){
            noObdClient.disconnect();
        }
        if(obdClient != null){
            obdClient.disconnect();
        }
        if(personalClient != null){
            personalClient.disconnect();
        }
    }


    private void dealPersonalAssetMessage(Message message) throws IOException {
        if (message instanceof SignInMessage){
            SignInMessage signInMessage = (SignInMessage)message;
            System.out.println("receive sign in message :" + signInMessage.getImei());
            byte[] reply = personalAssetMsgEncoder.getSignInMsgReply(signInMessage.getImei(), true, signInMessage.getSerialNo());
            personalClient.publicMessage(reply);
        }else if (message instanceof HeartbeatMessage){
            HeartbeatMessage heartbeatMessage = (HeartbeatMessage)message;
            System.out.println("receive heartbeat message :" + heartbeatMessage.getImei());
            byte[] reply = personalAssetMsgEncoder.getHeartbeatMsgReply(heartbeatMessage.getImei(), true, heartbeatMessage.getSerialNo());
            personalClient.publicMessage(reply);
        }else if (message instanceof LocationInfoMessage){
            LocationInfoMessage locationInfoMessage = (LocationInfoMessage)message;
            System.out.println("receive location info message :" + locationInfoMessage.getImei());

            byte[] reply = personalAssetMsgEncoder.getLocationMsgReply(locationInfoMessage.getImei(), true, locationInfoMessage.getSerialNo(),locationInfoMessage.getProtocolHeadType());
            personalClient.publicMessage(reply);
        }else if (message instanceof LocationAlarmMessage){
            LocationAlarmMessage locationAlarmMessage = (LocationAlarmMessage)message;
            System.out.println("receive location alarm message :" + locationAlarmMessage.getImei() + " Alarm is : " + locationAlarmMessage.getOriginalAlarmCode());
            byte[] reply = personalAssetMsgEncoder.getLocationAlarmMsgReply(locationAlarmMessage.getImei(),true, locationAlarmMessage.getSerialNo(),locationAlarmMessage.getOriginalAlarmCode(), locationAlarmMessage.getProtocolHeadType());
            personalClient.publicMessage(reply);
        }else if (message instanceof ConfigMessage){
            ConfigMessage configMessage = (ConfigMessage)message;
            System.out.println("receive config message :" + configMessage.getImei() + " : " + configMessage.getConfigResultContent());
        }else if (message instanceof ForwardMessage){
            ForwardMessage forwardMessage = (ForwardMessage)message;
            System.out.println("receive forward message :" + forwardMessage.getImei() + " : " + forwardMessage.getContent());
        }else if (message instanceof USSDMessage){
            USSDMessage ussdMessage = (USSDMessage)message;
            System.out.println("receive forward message :" + ussdMessage.getImei() + " : " + ussdMessage.getContent());
        }else if (message instanceof RS232Message){
            RS232Message rs232Message = (RS232Message)message;
            System.out.println("receive RS232 message :" + rs232Message.getImei());
        }else if (message instanceof NetworkInfoMessage){
            NetworkInfoMessage NetworkInfoMessage = (NetworkInfoMessage)message;
            System.out.println("receive network info message :" + NetworkInfoMessage.getImei());
        }else if (message instanceof WifiMessage){
            WifiMessage wifiMessage = (WifiMessage)message;
            System.out.println("receive wifi message :" + wifiMessage.getImei());
            byte[] reply = personalAssetMsgEncoder.getWifiMsgReply(wifiMessage.getImei(), true, wifiMessage.getSerialNo());
            personalClient.publicMessage(reply);
        }else if (message instanceof LockMessage){
            LockMessage lockMessage = (LockMessage)message;
            System.out.println("receive lock message :" + lockMessage.getImei());
            byte[] reply = personalAssetMsgEncoder.getWifiMsgReply(lockMessage.getImei(), true, lockMessage.getSerialNo());
            personalClient.publicMessage(reply);
        }else if (message instanceof BluetoothPeripheralDataMessage){
            BluetoothPeripheralDataMessage bluetoothPeripheralDataMessage = (BluetoothPeripheralDataMessage)message;
            System.out.println("receive bluetooth ignition message :" + bluetoothPeripheralDataMessage.getImei());
            byte[] reply = personalAssetMsgEncoder.getBluetoothPeripheralMsgReply(bluetoothPeripheralDataMessage.getImei(), bluetoothPeripheralDataMessage.getSerialNo(),bluetoothPeripheralDataMessage.getProtocolHeadType());
            personalClient.publicMessage(reply);
        }
    }

    private void dealObdMessage(Message message) throws IOException {
        if (message instanceof SignInMessage){
            SignInMessage signInMessage = (SignInMessage)message;
            System.out.println("receive sign in message :" + signInMessage.getImei());
            byte[] reply = t880xdEncoder.getSignInMsgReply(signInMessage.getImei(), true, signInMessage.getSerialNo());
            obdClient.publicMessage(reply);
        }else if (message instanceof HeartbeatMessage){
            HeartbeatMessage heartbeatMessage = (HeartbeatMessage)message;
            System.out.println("receive heartbeat message :" + heartbeatMessage.getImei());
            byte[] reply = t880xdEncoder.getHeartbeatMsgReply(heartbeatMessage.getImei(), true, heartbeatMessage.getSerialNo());
            obdClient.publicMessage(reply);
        }else if (message instanceof LocationInfoMessage){
            LocationInfoMessage locationInfoMessage = (LocationInfoMessage)message;
            System.out.println("receive location info message :" + locationInfoMessage.getImei());
            byte[] reply = t880xdEncoder.getLocationMsgReply(locationInfoMessage.getImei(), true, locationInfoMessage.getSerialNo(), locationInfoMessage.getProtocolHeadType());
            obdClient.publicMessage(reply);
        }else if (message instanceof LocationAlarmMessage){
            LocationAlarmMessage locationAlarmMessage = (LocationAlarmMessage)message;
            System.out.println("receive location alarm message :" + locationAlarmMessage.getImei() + " Alarm is : "+ locationAlarmMessage.getOriginalAlarmCode());
            byte[] reply = t880xdEncoder.getLocationAlarmMsgReply(locationAlarmMessage.getImei(),true, locationAlarmMessage.getSerialNo(),locationAlarmMessage.getOriginalAlarmCode(),locationAlarmMessage.getProtocolHeadType());
            obdClient.publicMessage(reply);
        }else if(message instanceof  GpsDriverBehaviorMessage){
            GpsDriverBehaviorMessage gpsDriverBehaviorMessage = (GpsDriverBehaviorMessage)message;
            System.out.println("receive gps driver behavior message :" + gpsDriverBehaviorMessage.getImei());
            System.out.println("behavior is :" + getGpsDriverBehaviorDescription(gpsDriverBehaviorMessage.getBehaviorType()));
            byte[] reply = t880xdEncoder.getGpsDriverBehaviorMsgReply(gpsDriverBehaviorMessage.getImei(), gpsDriverBehaviorMessage.getSerialNo());
            obdClient.publicMessage(reply);

        }else if (message instanceof AccelerationDriverBehaviorMessage){
            AccelerationDriverBehaviorMessage accelerationDriverBehaviorMessage = (AccelerationDriverBehaviorMessage)message;
            System.out.println("receive acceleration driver behavior message :" + accelerationDriverBehaviorMessage.getImei());
            System.out.println("behavior is :" + getGpsDriverBehaviorDescription(accelerationDriverBehaviorMessage.getBehaviorType()));
            byte[] reply = t880xdEncoder.getAccelerationDriverBehaviorMsgReply(accelerationDriverBehaviorMessage.getImei(), accelerationDriverBehaviorMessage.getSerialNo());
            obdClient.publicMessage(reply);
        }else if (message instanceof AccidentAccelerationMessage){
            AccidentAccelerationMessage accidentAccelerationMessage = (AccidentAccelerationMessage)message;
            System.out.println("receive accident acceleration message :" + accidentAccelerationMessage.getImei());
            byte[] reply = t880xdEncoder.getAccelerationAlarmMsgReply(accidentAccelerationMessage.getImei(), accidentAccelerationMessage.getSerialNo());
            obdClient.publicMessage(reply);
        }else if (message instanceof BluetoothPeripheralDataMessage){
            BluetoothPeripheralDataMessage bluetoothPeripheralDataMessage = (BluetoothPeripheralDataMessage)message;
            System.out.println("receive bluetooth ignition message :" + bluetoothPeripheralDataMessage.getImei());
            byte[] reply = t880xdEncoder.getBluetoothPeripheralMsgReply(bluetoothPeripheralDataMessage.getImei(), bluetoothPeripheralDataMessage.getSerialNo(),bluetoothPeripheralDataMessage.getProtocolHeadType());
            obdClient.publicMessage(reply);
        }else if (message instanceof ConfigMessage){
            ConfigMessage configMessage = (ConfigMessage)message;
            System.out.println("receive config message :" + configMessage.getImei() + " : " + configMessage.getConfigResultContent());
        }else if (message instanceof ForwardMessage){
            ForwardMessage forwardMessage = (ForwardMessage)message;
            System.out.println("receive forward message :" + forwardMessage.getImei() + " : " + forwardMessage.getContent());
        }else if (message instanceof USSDMessage){
            USSDMessage ussdMessage = (USSDMessage)message;
            System.out.println("receive forward message :" + ussdMessage.getImei() + " : " + ussdMessage.getContent());
        }else if (message instanceof RS232Message){
            RS232Message rs232Message = (RS232Message)message;
            System.out.println("receive RS232 message :" + rs232Message.getImei());
        }else if (message instanceof NetworkInfoMessage){
            NetworkInfoMessage NetworkInfoMessage = (NetworkInfoMessage)message;
            System.out.println("receive network info message :" + NetworkInfoMessage.getImei());
            byte[] reply = t880xdEncoder.getNetworkMsgReply(NetworkInfoMessage.getImei(), NetworkInfoMessage.getSerialNo());
            obdClient.publicMessage(reply);
        }
    }


    private void dealNoObdMessage(Message message) throws IOException {
        if (message instanceof SignInMessage){
            SignInMessage signInMessage = (SignInMessage)message;
            System.out.println("receive sign in message :" + signInMessage.getImei());
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getSignInMsgReply(signInMessage.getImei(), true, signInMessage.getSerialNo());
            noObdClient.publicMessage(reply);
        }else if (message instanceof HeartbeatMessage){
            HeartbeatMessage heartbeatMessage = (HeartbeatMessage)message;
            System.out.println("receive heartbeat message :" + heartbeatMessage.getImei());
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getHeartbeatMsgReply(heartbeatMessage.getImei(), true, heartbeatMessage.getSerialNo());
            noObdClient.publicMessage(reply);
        }else if (message instanceof LocationInfoMessage){
            LocationInfoMessage locationInfoMessage = (LocationInfoMessage)message;
            System.out.println("receive location info message :" + locationInfoMessage.getImei());
            //8806 Plus or some new model device, these is the code of 8806 plus.
            byte[] reply = t880xPlusEncoder.getLocationMsgReply(locationInfoMessage.getImei(), true, locationInfoMessage.getSerialNo(),message.getProtocolHeadType());
            noObdClient.publicMessage(reply);
        }else if (message instanceof LocationAlarmMessage){
            LocationAlarmMessage locationAlarmMessage = (LocationAlarmMessage)message;
            System.out.println("receive location alarm message :" + locationAlarmMessage.getImei() + " Alarm is : "+ locationAlarmMessage.getOriginalAlarmCode());
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getLocationAlarmMsgReply(locationAlarmMessage.getImei(),true, locationAlarmMessage.getSerialNo(),locationAlarmMessage.getOriginalAlarmCode(),locationAlarmMessage.getProtocolHeadType());
            noObdClient.publicMessage(reply);
        }else if(message instanceof  GpsDriverBehaviorMessage){
            GpsDriverBehaviorMessage gpsDriverBehaviorMessage = (GpsDriverBehaviorMessage)message;
            System.out.println("receive gps driver behavior message :" + gpsDriverBehaviorMessage.getImei());
            System.out.println("behavior is :" + getGpsDriverBehaviorDescription(gpsDriverBehaviorMessage.getBehaviorType()));
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getGpsDriverBehaviorMsgReply(gpsDriverBehaviorMessage.getImei(), gpsDriverBehaviorMessage.getSerialNo());
            noObdClient.publicMessage(reply);

        }else if (message instanceof AccelerationDriverBehaviorMessage){
            AccelerationDriverBehaviorMessage accelerationDriverBehaviorMessage = (AccelerationDriverBehaviorMessage)message;
            System.out.println("receive acceleration driver behavior message :" + accelerationDriverBehaviorMessage.getImei());
            System.out.println("behavior is :" + getGpsDriverBehaviorDescription(accelerationDriverBehaviorMessage.getBehaviorType()));
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getAccelerationDriverBehaviorMsgReply(accelerationDriverBehaviorMessage.getImei(), accelerationDriverBehaviorMessage.getSerialNo());
            noObdClient.publicMessage(reply);
        }else if (message instanceof AccidentAccelerationMessage){
            AccidentAccelerationMessage accidentAccelerationMessage = (AccidentAccelerationMessage)message;
            System.out.println("receive accident acceleration message :" + accidentAccelerationMessage.getImei());
            //8806 Plus or some new model device,need serial no,reply this message
            byte[] reply = t880xPlusEncoder.getAccelerationAlarmMsgReply(accidentAccelerationMessage.getImei(), accidentAccelerationMessage.getSerialNo());
            noObdClient.publicMessage(reply);
        }else if (message instanceof BluetoothPeripheralDataMessage){
            BluetoothPeripheralDataMessage bluetoothPeripheralDataMessage = (BluetoothPeripheralDataMessage)message;
            System.out.println("receive bluetooth ignition message :" + bluetoothPeripheralDataMessage.getImei());
            byte[] reply = t880xPlusEncoder.getBluetoothPeripheralMsgReply(bluetoothPeripheralDataMessage.getImei(), bluetoothPeripheralDataMessage.getSerialNo(),bluetoothPeripheralDataMessage.getProtocolHeadType());
            noObdClient.publicMessage(reply);
        }else if (message instanceof ConfigMessage){
            ConfigMessage configMessage = (ConfigMessage)message;
            System.out.println("receive config message :" + configMessage.getImei() + " : " + configMessage.getConfigResultContent());
        }else if (message instanceof ForwardMessage){
            ForwardMessage forwardMessage = (ForwardMessage)message;
            System.out.println("receive forward message :" + forwardMessage.getImei() + " : " + forwardMessage.getContent());
        }else if (message instanceof USSDMessage){
            USSDMessage ussdMessage = (USSDMessage)message;
            System.out.println("receive forward message :" + ussdMessage.getImei() + " : " + ussdMessage.getContent());
        }else if (message instanceof RS232Message){
            RS232Message rs232Message = (RS232Message)message;
            System.out.println("receive RS232 message :" + rs232Message.getImei());
        }else if (message instanceof NetworkInfoMessage){
            NetworkInfoMessage NetworkInfoMessage = (NetworkInfoMessage)message;
            System.out.println("receive network info message :" + NetworkInfoMessage.getImei());
            byte[] reply = t880xPlusEncoder.getNetworkMsgReply(NetworkInfoMessage.getImei(), NetworkInfoMessage.getSerialNo());
            noObdClient.publicMessage(reply);
        }
    }

    public String getAccelerationDriverBehaviorDescription(int behaviorType){
        if (behaviorType == AccelerationDriverBehaviorMessage.BEHAVIOR_ACCELERATE){
            return "The vehicle bad acceleration";
        }else if (behaviorType == AccelerationDriverBehaviorMessage.BEHAVIOR_TURN_AND_BRAKE){
            return "The vehicle brake";
        }else {
            return "";
        }
    }

    public String getGpsDriverBehaviorDescription(int behaviorType){
        if (behaviorType == GpsDriverBehaviorType.HIGH_SPEED_ACCELERATE){
            return "The vehicle accelerates at the high speed.";
        }else if (behaviorType == GpsDriverBehaviorType.HIGH_SPEED_BRAKE){
            return "The vehicle brakes  at the high speed.";
        }else if (behaviorType == GpsDriverBehaviorType.MEDIUM_SPEED_ACCELERATE){
            return "The vehicle accelerates at the high speed.";
        }else if (behaviorType == GpsDriverBehaviorType.MEDIUM_SPEED_ACCELERATE){
            return "The vehicle brakes  at the high speed.";
        }else if (behaviorType == GpsDriverBehaviorType.LOW_SPEED_ACCELERATE){
            return "The vehicle accelerates at the high speed.";
        }else if (behaviorType == GpsDriverBehaviorType.LOW_SPEED_ACCELERATE){
            return "The vehicle brakes  at the high speed.";
        }else {
            return "";
        }
    }
}

File: pythonTopflytechCodec-3.9/MqttExample.py
from TopflytechCodec import *
from MqttClient import *



def getGpsDriverBehaviorDescription(behaviorType):
    if behaviorType == GpsDriverBehaviorType.HIGH_SPEED_ACCELERATE:
        return "The vehicle accelerates at the high speed."
    elif behaviorType == GpsDriverBehaviorType.HIGH_SPEED_BRAKE:
        return "The vehicle brakes  at the high speed."
    elif behaviorType == GpsDriverBehaviorType.MEDIUM_SPEED_ACCELERATE:
        return "The vehicle accelerates at the high speed."
    elif behaviorType == GpsDriverBehaviorType.MEDIUM_SPEED_ACCELERATE:
        return "The vehicle brakes  at the high speed."
    elif behaviorType == GpsDriverBehaviorType.LOW_SPEED_ACCELERATE:
        return "The vehicle accelerates at the high speed."
    elif behaviorType == GpsDriverBehaviorType.LOW_SPEED_ACCELERATE:
        return "The vehicle brakes  at the high speed."
    else:
        return ""



t880xPlusEncoder = T880xPlusEncoder(MessageEncryptType.NONE,"")
t880xdEncoder = T880xdEncoder(MessageEncryptType.NONE,"")
personalEncoder = PersonalAssetMsgEncoder(MessageEncryptType.NONE,"")

def dealObdDeviceMessage(message,socketClient):
    if isinstance(message,SignInMessage):
        print ("receive signInMessage: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,HeartbeatMessage):
        print ("receive heartbeatMessage" + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,LocationInfoMessage):
        print ("receive locationInfoMessage" + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getLocationMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getLocationMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,LocationAlarmMessage):
        print ("receive locationAlarmMessage" + message.imei + " Alarm is : " + str(message.originalAlarmCode))
        # some new model device,need serial no,reply this message
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,GpsDriverBehaviorMessage):
        print ("receive gpsDriverBehaviorMessage" + message.imei + " behavior is :" + getGpsDriverBehaviorDescription(message.behaviorType))
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getGpsDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getGpsDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,AccelerationDriverBehaviorMessage):
        print ("receive accelerationDriverBehaviorMessage" + message.imei + " behavior is :" + getGpsDriverBehaviorDescription(message.behaviorType))
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getAccelerationDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getAccelerationDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,AccidentAccelerationMessage):
        print ("receive accidentAccelerationMessage" + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getAccelerationAlarmMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getAccelerationAlarmMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,ConfigMessage):
        print ("receive configMessage: " + message.imei + " : " + message.configContent)
    elif isinstance(message,ForwardMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,USSDMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,ObdMessage):
        print ("receive OBD Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getObdMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getObdMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,WifiMessage):
        print ("receive WIFI Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getObdMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getWifiMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,BluetoothPeripheralDataMessage):
        print ("receive blue Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,NetworkInfoMessage):
        print ("receive network info Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xdEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xdEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)


def dealNoObdDeviceMessage(message,socketClient):
    """
    Device model like :8806 plus, use this method.
    :param message:
    :param socketClient:
    :return:
    """
    if isinstance(message,SignInMessage):
        print ("receive signInMessage: " + message.imei)
        #8806 Plus or some new model device,need serial no,reply this message
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,HeartbeatMessage):
        print ("receive heartbeatMessage" + message.imei)
        #8806 Plus or some new model device,need serial no,reply this message
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,LocationInfoMessage):
        print ("receive locationInfoMessage" + message.imei)
        #8806 Plus or some new model device, these is the code of 8806 plus.
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getLocationMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getLocationMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,LocationAlarmMessage):
        print ("receive locationAlarmMessage" + message.imei + "Alarm is : " + str(message.originalAlarmCode))
        #8806 Plus or some new model device,need serial no,reply this message
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,GpsDriverBehaviorMessage):
        print ("receive gpsDriverBehaviorMessage" + message.imei + " behavior is :" + getGpsDriverBehaviorDescription(message.behaviorType))
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getGpsDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getGpsDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,AccelerationDriverBehaviorMessage):
        print ("receive accelerationDriverBehaviorMessage" + message.imei + " behavior is :" + getGpsDriverBehaviorDescription(message.behaviorType))
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getAccelerationDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getAccelerationDriverBehaviorMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,AccidentAccelerationMessage):
        print ("receive accidentAccelerationMessage" + message.imei)
        #8806 Plus or some new model device,need serial no,reply this message
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getAccelerationAlarmMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getAccelerationAlarmMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,ConfigMessage):
        print ("receive configMessage: " + message.imei + " : " + message.configContent)
    elif isinstance(message,ForwardMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,USSDMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,RS232Message):
        print ("receive RS232 Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getRS232MsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getRS232MsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,BluetoothPeripheralDataMessage):
        print ("receive blue Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,NetworkInfoMessage):
        print ("receive network info Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = t880xPlusEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = t880xPlusEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,WifiMessage):
        print ("receive wifi Message: " + message.imei)
        reply = t880xPlusEncoder.getWifiMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,OneWireMessage):
        print ("receive one wire Message: " + message.imei)
        reply = t880xPlusEncoder.getOneWireMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,RS485Message):
        print ("receive rs485 Message: " + message.imei)
        reply = t880xPlusEncoder.getRs485MsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,ObdMessage):
        print ("receive DTC(OBD) Message: " + message.imei)
        reply = t880xPlusEncoder.getObdMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)

def dealPersonalDeviceMessage(message,socketClient):
    """
    Device model like :8806 plus, use this method.
    :param message:
    :param socketClient:
    :return:
    """
    if isinstance(message,SignInMessage):
        print ("receive signInMessage: " + message.imei)
        # if message.isNeedResp:
        #     reply = personalEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = personalEncoder.getSignInMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,HeartbeatMessage):
        print ("receive heartbeatMessage" + message.imei)
        # if message.isNeedResp:
        #     reply = personalEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = personalEncoder.getHeartbeatMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,LocationInfoMessage):
        print ("receive locationInfoMessage" + message.imei)
        # if message.isNeedResp:
        #     reply = personalEncoder.getLocationMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode)
        #     socketClient.publish(reply)
        reply = personalEncoder.getLocationMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,LocationAlarmMessage):
        print ("receive locationAlarmMessage" + message.imei + " Alarm is : " + str(message.originalAlarmCode))
        # if message.isNeedResp:
        #     reply = personalEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode)
        #     socketClient.publish(reply)
        reply = personalEncoder.getLocationAlarmMsgReply(message.imei,True,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,ConfigMessage):
        print ("receive configMessage: " + message.imei + " : " + message.configContent)
    elif isinstance(message,ForwardMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,USSDMessage):
        print ("receive forwardMessage: " + message.imei + " : " + message.content)
    elif isinstance(message,BluetoothPeripheralDataMessage):
        print ("receive blue Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = personalEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = personalEncoder.getBluetoothPeripheralMsgReply(message.imei,True,message.serialNo,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,WifiMessage):
        print ("receive wifi location Message: " + message.imei)
        reply = personalEncoder.getWifiMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,InnerGeoDataMessage):
        print ("receive inner geo data Message: " + message.imei)
        reply = personalEncoder.getInnerGeoDataMsgReply(message.imei,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message, DeviceTempCollectionMessage):
        print ("receive device extra temp data Message: " + message.imei)
        reply = personalEncoder.getDeviceTempCollectionMsgReply(message.imei, message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,WifiWithDeviceInfoMessage):
        print ("receive wifi with device info Message: " + message.imei)
        reply = personalEncoder.getWifiWithDeviceInfoMsgReply(message.imei,message.serialNo,message.originalAlarmCode,message.protocolHeadType)
        socketClient.publish(reply)
    elif isinstance(message,LockMessage):
        print ("receive lock Message: " + message.imei)
        reply = personalEncoder.getLockMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)
    elif isinstance(message,NetworkInfoMessage):
        print ("receive network info Message: " + message.imei)
        # if message.isNeedResp:
        #     reply = personalEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        #     socketClient.publish(reply)
        reply = personalEncoder.getNetworkMsgReply(message.imei,True,message.serialNo)
        socketClient.publish(reply)

decoder = Decoder(MessageEncryptType.NONE,"")
#decoder = ObdDecoder(MessageEncryptType.NONE,"")
#decoder = PersonalAssetMsgDecoder(MessageEncryptType.NONE,"")
if __name__ == "__main__":

    def custom_handle_message(message):
        messageList = decoder.decode(message)
        for message in messageList:
            dealNoObdDeviceMessage(message,client)
            # dealObdDeviceMessage(message,client)
            #dealPersonalDeviceMessage(message, client)


    imei = "518000518001517"

    client = MqttClient(imei, custom_handle_message)
    client.connect()
    import time

    time.sleep(200)
    client.disconnect()